### PR TITLE
fix #25: Add API token support

### LINKED ISSUE #25: Add support for API Tokens
The API now supports the use of API tokens as credentials (see https://api.cloudflare.com/#getting-started-requests). We should provide this option in the client.

### PR DESCRIPTION
**Pull request by gabbifish**
This PR adds support for api token usage (https://api.cloudflare.com/#getting-started-requests)

### CODE REVIEW

**Comment by gabbifish** [comment]
I updated the PR with the suggestions above. Please let me know what you think about the default option on the enum!

File: src/framework/auth.rs [hunk 1]
```diff
@@ -13,10 +15,14 @@ pub trait AuthClient {
...
     fn auth(self, credentials: &Credentials) -> RequestBuilder {
         match credentials {
-            Credentials::User { key, email } => self
-                .header("X-Auth-Key", key.as_str())
-                .header("X-Auth-Email", email.as_str()),
+            Credentials::UserAuthKey { email, key } => self
+                .header("X-Auth-Email", email.as_str())
+                .header("X-Auth-Key", key.clone()),
+            Credentials::UserAuthToken { email, token } => self
+                .header("X-Auth-Email", email.as_str())
```
Comment (ashleymichal) [inline src/framework/auth.rs]: according to the API docs, tokens are used by themselves, not in tandem with email

File: example/main.rs [hunk 2]
```diff
@@ -165,13 +171,28 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
...
+    let key = matches.value_of("auth-key");
+    let token = matches.value_of("auth-token");
 
-    let api_client = HttpApiClient::new(Credentials::User {
-        key: key.to_string(),
-        email: email.to_string(),
-    });
+    // Create bogus default api_client; this will be overwritten with a client
+    // that uses a key or token.
+    let mut credentials = Credentials::Default;
```
Comment (adamchalmers) [inline example/main.rs]: I don't think we need to add a special `Default` variant just for this assignment. Instead, we can assign `credentials` to the result of an `if`, like this:

```rust
let credentials = if let Some(key) = key {
    Credentials::UserAuthKey {
        email: email.unwrap().to_string(),
        key: key.to_string(),
    }
} else if let Some(token) = token {
    Credentials::UserAuthToken {
        token: token.to_string(),
    }
} else {
    panic!("either token or key must be specified")
}
```

File: src/framework/auth.rs [hunk 3]
```diff
@@ -2,8 +2,10 @@ use reqwest::RequestBuilder;
 
 #[derive(Debug)]
 pub enum Credentials {
-    User { key: String, email: String },
+    UserAuthKey { email: String, key: String },
+    UserAuthToken { token: String },
     Service { key: String },
+    Default,
```
Comment (adamchalmers) [inline src/framework/auth.rs]: See my comment on `example/main.rs` - I don't think we should add a `Default` variant. We don't ever want a user to actually use it, so let's just never implement it in the first place :)

File: src/framework/auth.rs [hunk 4]
```diff
@@ -13,10 +15,14 @@ pub trait AuthClient {
...
-                .header("X-Auth-Key", key.as_str())
-                .header("X-Auth-Email", email.as_str()),
+            Credentials::UserAuthKey { email, key } => self
+                .header("X-Auth-Email", email.as_str())
+                .header("X-Auth-Key", key.clone()),
+            Credentials::UserAuthToken { token } => {
+                self.header("Authorization", &format!("Bearer {}", token.clone()))
+            }
             Credentials::Service { key } => self.header("X-Auth-User-Service-Key", key.as_str()),
+            _ => self,
```
Comment (adamchalmers) [inline src/framework/auth.rs]: If you remove the `Default` variant, don't forget to remove this too.

File: example/main.rs (hunk 2 above)
Comment (gabbifish) [reply example/main.rs]: Ah! I'll try this. I tried using other error handling approaches (namely, failure crate) but the compiler was unhappy with me; just using `panic!` should work though!

File: src/framework/auth.rs (hunk 3 above)
Comment (gabbifish) [reply src/framework/auth.rs]: Agreed, it was more of a perceived necessary evil at the time :)

File: example/main.rs (hunk 2 above)
Comment (adamchalmers) [reply example/main.rs]: Great. I'd never put a `panic!` in the library itself, but in the example binary it's totally fine.

**Pull request merged**